api/v1beta1/cluster_types.go

```diff
@@ -451,6 +469,77 @@ type ClusterStatus struct {
 	// ObservedGeneration is the latest generation observed by the controller.
 	// +optional
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// v1beta2 groups all the fields that will be added or modified in Cluster's status with the V1Beta2 version.
+	// +optional
+	V1Beta2 *ClusterV1Beta2Status `json:"v1beta2,omitempty"`
+}
+
+// ClusterV1Beta2Status groups all the fields that will be added or modified in Cluster with the V1Beta2 version.
+// See https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240916-improve-status-in-CAPI-resources.md#machineset-newconditions for more context.
+type ClusterV1Beta2Status struct {
+	// conditions represents the observations of a Cluster's current state.
+	// Known condition types are Available, InfrastructureReady, ControlPlaneInitialized, ControlPlaneAvailable, WorkersAvailable, MachinesReady
+	// MachinesUpToDate, RemoteConnectionProbe, ScalingUp, ScalingDown, Remediating, Deleting, Paused.
+	// Additionally, a TopologyReconciled will be added in case the Cluster is referencing a ClusterClass / defining a managed Topology.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
Can we set a maximum length on this list? Perhaps 32 to give plenty of headroom?

Reply:
In general, you want to limit the amount of data that can be added to the API to prevent people just sending garbage.

Now in the case of this example, since it's status, we probably don't expect folks to add crazy length lists here. However, if in the future we decide we want to add some CEL validation that iterates over the conditions list, having a length limit will significantly reduce the runtime cost estimated.

When you don't have a length limit, the cardinality is unbounded and any validation within is multipled by the number of times you can fit the minimally serialised struct/field into the maximum storage size in etcd.

So it easily calculates cardinality in the millions when you have no length limit.

Adding a length limit later is harder than adding it now.

Reply:
CC @vincepri @sbueringer you might find the above conversation interesting

Reply:
Good points. When I was experimenting with CEL for ClusterClass variables the result was that without maxLength and friends CEL rules were just rejected by the apiserver because of ~ unbound runtime estimations

Reply:
Fixed!